walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,48 +13,84 @@
         }
     }
 
-    Collections.sort(unavailableTimes, TimeRange.ORDER_BY_START);
+    Collections.sort(unavailableTimesForMandatory, TimeRange.ORDER_BY_START);
+    Collections.sort(unavailableTimesForOptional, TimeRange.ORDER_BY_START);
 
-    if (unavailableTimes.size() > 0) {
-        combineOverlaps(unavailableTimes, unavailableTimes.get(0), 0);
+    if (unavailableTimesForMandatory.size() > 0) { // Do this only for mandatory attendees since only one optional attendee exists
+        combineOverlaps(unavailableTimesForMandatory, unavailableTimesForMandatory.get(0), 0);
     }
 
-    int start = 0;
-    for (TimeRange range : unavailableTimes) {
-        TimeRange availableRange = TimeRange.fromStartEnd(start, range.start(), false);
+    findAvailableRanges(unavailableTimesForMandatory, availableTimesForMandatory, request);
+    findAvailableRanges(unavailableTimesForOptional, availableTimesForOptional, request);
+
+    List<TimeRange> finalAvailableTimes = new ArrayList<TimeRange>();
+
+    if (request.getOptionalAttendees().size() == 0) {
+        finalAvailableTimes = availableTimesForMandatory;
+    } else if (request.getAttendees().size() == 0) {
+        finalAvailableTimes = availableTimesForOptional; 
+    } else if (availableTimesForMandatory.size() == 0) {
+        finalAvailableTimes = availableTimesForOptional;
+    } else if (availableTimesForOptional.size() == 0) { 
+        finalAvailableTimes = availableTimesForMandatory;
+    } else {
+        for (TimeRange m : availableTimesForMandatory) {
+            for (TimeRange o : availableTimesForOptional) {
+                TimeRange encompassed = null;
+    
+                if (m.start() >= o.start() && m.end() <= o.end()) {
+                    encompassed = m;
+                } else if (o.start() >= m.start() && o.end() <= m.end()) {
+                    encompassed = o;
+                }
+
+                if (encompassed != null) {
+                    finalAvailableTimes.add(encompassed);
+                }               
+            }   
+        }
 
-        addAvailableRanges(availableTimes, availableRange, request);
-        
-        start = range.end();
+        if (finalAvailableTimes.size() == 0) {
+            finalAvailableTimes = availableTimesForMandatory;
+        }
     }
 
-    TimeRange lastAvailableRange = TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY+1, false);
-
-    addAvailableRanges(availableTimes, lastAvailableRange, request);
-    
-    return availableTimes;   
+    return finalAvailableTimes;   
   }
  
-  public boolean attendeesFree(Event e, MeetingRequest request){
+  public boolean mandatoryAttendeesFree(Event e, MeetingRequest request){
     return Collections.disjoint(e.getAttendees(), request.getAttendees());
   }
 
+  public boolean optionalAttendeesFree(Event e, MeetingRequest request){
+    return Collections.disjoint(e.getAttendees(), request.getOptionalAttendees());
+  }
```

Comment:
Do these need to be separate functions? It seems like we could parametrize the attendees part :)

Generally, I like to follow the principle of DRY (don't repeat yourself) if possible. It generally helps reduce duplicate code.

Reply:
I'll keep that in mind from now on, thank you for the advice and thank you for approving the PR :)